fuel/datasets/base.py

```diff
@@ -177,88 +190,136 @@ def filter_sources(self, data):
-    ...     cPickle.dump(mnist, f)
-
-    You will notice that the dumping of the dataset was relatively quick,
-    because it didn't attempt to write MNIST to disk. We can now reload it,
-    and if the data file has not been moved, it will be as if nothing
-    happened.
-
-    >>> with open('mnist.pkl', 'rb') as f:
-    ...     mnist = cPickle.load(f)
-    >>> print(mnist.features.shape)
-    (60000, 784)
-
-    However, if the data files can't be found on disk, accessing the data
-    will fail.
-
-    >>> from fuel import config
-    >>> correct_path = config.data_path
-    >>> config.data_path = '/non/existing/path'
-    >>> with open('mnist.pkl', 'rb') as f:
-    ...     mnist = cPickle.load(f)
-    >>> print(mnist.features.shape) # doctest: +SKIP
-    Traceback (most recent call last):
-      ...
-    FileNotFoundError: [Errno 2] No such file or directory: ...
-
-    Because the loading happens lazily, we can still deserialize our
-    dataset, correct the situation, and then continue.
-
-    >>> config.data_path = correct_path
-    >>> print(mnist.features.shape)
-    (60000, 784)
-
-    .. doctest::
-       :hide:
-
-       >>> import os
-       >>> os.remove('mnist.pkl')
+    Internally, this method uses `picklable iterools's`_ ``_iter``
+    function, providing picklable alternatives to some iterators such as
+    :func:`range`, :func:`tuple`, and even :class:`file`. However, if the
+    iterable returns a different kind of iterator that is not picklable,
+    you might want to consider using the :func:`.do_not_pickle_attributes`
+    decorator.
 
+    To iterate over a container in batches, combine this dataset with the
+    :class:`BatchDataStream` data stream.
 
     """
-    def load(self):
-        """Load data from e.g. the file system.
+    example_iteration_scheme = None
+
+    def __init__(self, iterables, **kwargs):
+        if isinstance(iterables, dict):
+            self.provides_sources = tuple(iterables.keys())
+        else:
+            self.provides_sources = ('data',)
+        super(IterableDataset, self).__init__(**kwargs)
+        if isinstance(iterables, dict):
+            if not all(isinstance(iterable, collections.Iterable)
+                       for iterable in iterables.values()):
+                raise ValueError
+            self.iterables = [iterables[source] for source in self.sources]
+        else:
+            if not isinstance(iterables, collections.Iterable):
+                raise ValueError
+            self.iterables = [iterables]
 
-        Any interaction with the outside world e.g. the file system,
-        database connections, servers, etc. should be done in this method.
-        This allows datasets to be pickled and unpickled, even in
-        environments where the original data is unavailable or has changed
-        position.
+    @property
+    def num_examples(self):
+        try:
+            return min(len(iterable) for iterable in self.iterables)
```

Comment:
Yes, that particular case could be intercepted before iteration even begins. My suggestion would only be able to raise an error at the end of the first epoch, so maybe it's better to just silently ignore that and only check at the beginning.